src/custom/utils/tip.test.ts

```diff
@@ -0,0 +1,268 @@
+  outputCurrency: Currency
+  pair: Pair
+}
+
+function setupTrade(
+  props: SetupTrade
+):
+  | (SetupTrade & {
+      trade: Trade & {
+        inputAmountWithFees?: CurrencyAmount
+      }
+      typedAmountAsCurrency?: CurrencyAmount
+      feeAsCurrency?: CurrencyAmount
+    })
+  | null {
+  const {
+    type,
+    fee: { minimalFee = '10' },
+    typedAmount,
+    inputCurrency,
+    outputCurrency,
+    pair
+  } = props
+
+  // Trade type: IN // OUT
+  const isExactIn = type === TradeType.EXACT_INPUT
+
+  // Amounts in Currency form
+  const typedAmountAsCurrency = tryParseAmount(typedAmount, isExactIn ? inputCurrency : outputCurrency)
+  const feeAsCurrency = tryParseAmount(minimalFee, inputCurrency)
+
+  // Type checks mainly
+  if (!typedAmountAsCurrency || !pair || !feeAsCurrency) return null
+
+  // Calculate IN // OUT trades
+  let trade: Trade & {
+    inputAmountWithFees?: CurrencyAmount
+  }
+  if (isExactIn) {
+    // subtract fee from typed in amount
+    const typedAmountWithFees = typedAmountAsCurrency.subtract(feeAsCurrency)
+    const preTrade =
+      // This is how Uni calculates trades in their hook and in Trades.ts
+      Trade.bestTradeExactIn(
+        [pair1and2],
+        // subtract fee from typed in amount
+        typedAmountWithFees,
+        outputCurrency,
+        DEFAULT_HOPS
+        // Yes, this is also copy/paste
+      )[0] ?? null
+
+    // We need to iverride the Trade object to use different values as we are intercepting initial inputs
+    // and applying fee. For ExactIn orders, we leave outputAmount as is
+    // and only change inputAmount to show the original entry before fee calculation
+    trade = {
+      ...preTrade,
+      get inputAmount() {
+        return typedAmountAsCurrency!
+      },
+      get minimumAmountOut() {
+        return this.minimumAmountOut
+      },
+      get maximumAmountIn() {
+        return this.maximumAmountIn
+      },
+      inputAmountWithFees: typedAmountWithFees
+    }
+  } else {
+    // This is how Uni calculates trades in their hook and in Trades.ts
+    // Yes, the ?? null is also copy/paste
+    const preTrade = Trade.bestTradeExactOut([pair1and2], inputCurrency, typedAmountAsCurrency, DEFAULT_HOPS)[0] ?? null
+
+    const inputAmountWithFee = preTrade.inputAmount!.add(feeAsCurrency!)
+    // We need to iverride the Trade object to use different values as we are intercepting initial inputs
+    // and applying fee. For ExactOut orders, we leave inputAmount as is
+    // and only change outputAm to show the original entry before fee calculation
+    trade = {
+      ...preTrade,
+      inputAmount: inputAmountWithFee,
+      get minimumAmountOut() {
+        return this.minimumAmountOut
+      },
+      get maximumAmountIn() {
+        return this.maximumAmountIn
+      }
+    }
+  }
+
+  return {
+    ...props,
+    trade,
+    typedAmountAsCurrency,
+    feeAsCurrency
+  }
+}
+
+// CONFIG
+const typedAmount = (Math.random() * 9382).toString()
```

Comment:
unit tests with random stuff?

Reply:
:scream:

Reply:
well the tests are just: "make sure EXACT-IN and EXACT-OUT are the same. So what the numbers are didn't seem to be important as long as the math matched. Will change this